roachtest: scaledata/distributed_semaphore/nodes=3 failed

SHA: https://github.com/cockroachdb/cockroach/commits/a200cea4368ec90aaee12337d7ab5f9ca555108f

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=scaledata/distributed_semaphore/nodes=3 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1191939&tab=buildLog

```
The test failed on master:
	cluster.go:1626,scaledata.go:126,scaledata.go:53,test.go:1214: unexpected node event: 1: dead

```

```
panic: runtime error: slice bounds out of range [recovered]
	panic: panic while executing 1 statements: SELECT * FROM _ WHERE (_._ IS _) AND ((_ = $1)); caused by runtime error: slice bounds out of range

goroutine 1723 [running]:
panic(0x31059a0, 0xc007a198f0)
	/usr/local/go/src/runtime/panic.go:556 +0x2cb fp=0xc00779f548 sp=0xc00779f4b8 pc=0x72b49b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc006c54880, 0x39fde00, 0xc00763e5c0, 0x2ea0e00, 0x56ca260)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:719 +0x36d fp=0xc00779f608 sp=0xc00779f548 pc=0x1fae87d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc006c54880, 0x39fde00, 0xc00763e5c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:431 +0x61 fp=0xc00779f640 sp=0xc00779f608 pc=0x213b321
runtime.call32(0x0, 0x33fa2f0, 0xc006ab0800, 0x1800000018)
	/usr/local/go/src/runtime/asm_amd64.s:522 +0x3b fp=0xc00779f670 sp=0xc00779f640 pc=0x75a06b
panic(0x2ea0e00, 0x56ca260)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9 fp=0xc00779f700 sp=0xc00779f670 pc=0x72b389
runtime.panicslice()
	/usr/local/go/src/runtime/panic.go:54 +0xfa fp=0xc00779f730 sp=0xc00779f700 pc=0x72a2fa
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*zigzagJoiner).produceSpanFromBaseRow(0xc0077f9800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/zigzagjoiner.go:610 +0x56f fp=0xc00779f928 sp=0xc00779f730 pc=0x1e0f65f
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*zigzagJoiner).nextRow(0xc0077f9800, 0x39fdec0, 0xc007a19200, 0xc0075b8990, 0xc0076646e0, 0x7, 0x7, 0x7)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/zigzagjoiner.go:747 +0x109 fp=0xc00779fb68 sp=0xc00779f928 pc=0x1e10469
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*zigzagJoiner).Next(0xc0077f9800, 0x309f400, 0x3156b20, 0x3a33c01, 0x3a33c20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/zigzagjoiner.go:926 +0xa4 fp=0xc00779fc60 sp=0xc00779fb68 pc=0x1e11e74
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*joinReader).readInput(0xc006941500, 0xc007966b08, 0x3a33c20)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/joinreader.go:328 +0x68 fp=0xc00779fe10 sp=0xc00779fc60 pc=0x1dd0498
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*joinReader).Next(0xc006941500, 0xc007a19200, 0x3297e1d, 0xb, 0x39fdec0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/joinreader.go:301 +0x2da fp=0xc00779fe68 sp=0xc00779fe10 pc=0x1dd03ca
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x39fdec0, 0xc007a19200, 0x3a09fc0, 0xc006941500, 0x39ef280, 0xc006c5e380)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:174 +0x35 fp=0xc00779fea8 sp=0xc00779fe68 pc=0x1dadb05
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc006941500, 0x39fdec0, 0xc007a19200)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:802 +0x92 fp=0xc00779fee8 sp=0xc00779fea8 pc=0x1ddead2
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc0073a65a0, 0x39fdec0, 0xc007a19200, 0x33fa270, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:627 +0x1e9 fp=0xc00779ff48 sp=0xc00779fee8 pc=0x1dbdfa9
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000029180, 0xc007230480, 0xc00755ca20, 0xc0077a0898, 0xc007864840, 0xc006c54c90, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:252 +0x8ad fp=0xc0077a0788 sp=0xc00779ff48 pc=0x2016acd
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000029180, 0x39fdec0, 0xc007a189f0, 0xc006c54c90, 0xc007230480, 0xc00755ca20, 0x39ffac0, 0xc00723cc60, 0xc007864840)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:840 +0x227 fp=0xc0077a0a48 sp=0xc0077a0788 pc=0x201a1d7
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc006c54880, 0x39fdec0, 0xc007a189f0, 0xc006c54bd0, 0x3, 0x7f0f30357360, 0xc0078b5340, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1121 +0x283 fp=0xc0077a0ad8 sp=0xc0077a0a48 pc=0x1fc0f63
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc006c54880, 0x39fdec0, 0xc007a189f0, 0xc006c54bd0, 0x7f0f30357360, 0xc0078b5340, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:957 +0x658 fp=0xc0077a0bd8 sp=0xc0077a0ad8 pc=0x1fbfe58
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc006c54880, 0x39fdec0, 0xc007a189f0, 0x3a03a00, 0xc007716280, 0xc0060f91ee, 0x5a, 0x1, 0xc0079d6240, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:459 +0xdf6 fp=0xc0077a13b8 sp=0xc0077a0bd8 pc=0x1fbbe16
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc006c54880, 0x39fdec0, 0xc007a189f0, 0x3a03a00, 0xc007716280, 0xc0060f91ee, 0x5a, 0x1, 0xc0079d6240, 0x7, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:101 +0x610 fp=0xc0077a1660 sp=0xc0077a13b8 pc=0x1fbaba0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc006c54880, 0x39fde00, 0xc00763e5c0, 0xc000bf8c38, 0x5400, 0x15000, 0xc000bf8cd0, 0xc00753f020, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1242 +0x1418 fp=0xc0077a1ec8 sp=0xc0077a1660 pc=0x1fb12a8
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000cb38c0, 0x39fde00, 0xc00763e5c0, 0xc006c54880, 0x5400, 0x15000, 0xc000bf8cd0, 0xc00753f020, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:433 +0xce fp=0xc0077a1f28 sp=0xc0077a1ec8 pc=0x1facbfe
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc000cb38c0, 0x39fde00, 0xc00763e5c0, 0xc007686900, 0xc00744b8c4, 0xc00753f020, 0xc006c54880, 0x5400, 0x15000, 0xc000bf8cd0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:338 +0xe6 fp=0xc0077a1f88 sp=0xc0077a1f28 pc=0x22337c6
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1333 +0x1 fp=0xc0077a1f90 sp=0xc0077a1f88 pc=0x75bd81
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:321 +0x1080
```

--- a/pkg/sql/opt/exec/execbuilder/testdata/join
+++ b/pkg/sql/opt/exec/execbuilder/testdata/join
@@ -1469,19 +1469,13 @@ CREATE TABLE zigzag2 (
 query TTT
 EXPLAIN SELECT * FROM zigzag2 WHERE a = 1 AND b = 2 AND c IS NULL
 ----
-render                 ·          ·
- └── lookup-join       ·          ·
-      │                table      zigzag2@primary
-      │                type       inner
-      └── zigzag-join  ·          ·
-           │           type       inner
-           │           pred       ((@1 = 1) AND (@2 = 2)) AND (@4 IS NULL)
-           ├── scan    ·          ·
-           │           table      zigzag2@a_b_idx
-           │           fixedvals  2 columns
-           └── scan    ·          ·
-·                      table      zigzag2@c_idx
-·                      fixedvals  1 column
+filter           ·       ·
+ │               filter  c IS NULL
+ └── index-join  ·       ·
+      │          table   zigzag2@primary
+      └── scan   ·       ·
+·                table   zigzag2@a_b_idx
+·                spans   /1/2-/1/3
 
 # Test that we can force a merge join.
 query TTT

--- a/pkg/sql/opt/xform/custom_funcs.go
+++ b/pkg/sql/opt/xform/custom_funcs.go
@@ -849,8 +849,15 @@ func eqColsForZigzag(
 ) (leftEqPrefix, rightEqPrefix opt.ColList) {
 	leftEqPrefix = make(opt.ColList, 0, len(leftEqCols))
 	rightEqPrefix = make(opt.ColList, 0, len(rightEqCols))
-	i, leftCnt := 0, leftIndex.ColumnCount()
-	j, rightCnt := 0, rightIndex.ColumnCount()
+	// We can only zigzag on columns present in the key component of the index,
+	// so use the LaxKeyColumnCount here because that's the longest prefix of the
+	// columns in the index which is guaranteed to exist in the key component.
+	// Using KeyColumnCount is invalid, because if we have a unique index with
+	// nullable columns, the "key columns" include the primary key of the table,
+	// which is only present in the key component if one of the other columns is
+	// NULL.
+	i, leftCnt := 0, leftIndex.LaxKeyColumnCount()
+	j, rightCnt := 0, rightIndex.LaxKeyColumnCount()
 	for ; i < leftCnt; i++ {
 		colID := tabID.ColumnID(leftIndex.Column(i).Ordinal)
 		if !fixedCols.Contains(int(colID)) {

--- a/pkg/sql/opt/xform/testdata/rules/join
+++ b/pkg/sql/opt/xform/testdata/rules/join
@@ -119,6 +119,28 @@ TABLE pqr
       ├── s string
       └── p int not null
 
+exec-ddl
+CREATE TABLE zz (
+    a INT8 PRIMARY KEY,
+    b INT8 NULL,
+    c INT8 NULL,
+    INDEX idx_b (b ASC),
+    CONSTRAINT idx_c UNIQUE (c)
+)
+----
+TABLE zz
+ ├── a int not null
+ ├── b int
+ ├── c int
+ ├── INDEX primary
+ │    └── a int not null
+ ├── INDEX idx_b
+ │    ├── b int
+ │    └── a int not null
+ └── INDEX idx_c
+      ├── c int
+      └── a int not null (storing)
+
 # --------------------------------------------------
 # CommuteJoin
 # --------------------------------------------------
@@ -1681,6 +1703,35 @@ inner-join (zigzag pqr@q pqr@s)
       ├── q = 1 [type=bool, outer=(2), constraints=(/2: [/1 - /1]; tight), fd=()-->(2)]
       ├── r = 1 [type=bool, outer=(3), constraints=(/3: [/1 - /1]; tight), fd=()-->(3)]
       └── s = 'foo' [type=bool, outer=(4), constraints=(/4: [/'foo' - /'foo']; tight), fd=()-->(4)]
+
+# Don't generate a zigzag which has the PK as its equality columns against
+# nullable unique indexes where the primary key is not part of the indexed
+# columns.
+
+# Regression test for #36051: prior to fixing this, we would try to use the PK
+# as the equality column here, but it's not actually part of the key so we
+# can't zigzag on it.
+opt
+SELECT * FROM zz WHERE b IS NULL AND c = 2
+----
+select
+ ├── columns: a:1(int!null) b:2(int) c:3(int!null)
+ ├── cardinality: [0 - 1]
+ ├── key: ()
+ ├── fd: ()-->(1-3)
+ ├── index-join zz
+ │    ├── columns: a:1(int!null) b:2(int) c:3(int)
+ │    ├── cardinality: [0 - 1]
+ │    ├── key: ()
+ │    ├── fd: ()-->(1-3)
+ │    └── scan zz@idx_c
+ │         ├── columns: a:1(int!null) c:3(int!null)
+ │         ├── constraint: /3: [/2 - /2]
+ │         ├── cardinality: [0 - 1]
+ │         ├── key: ()
+ │         └── fd: ()-->(1,3)
+ └── filters
+      └── b IS NULL [type=bool, outer=(2), constraints=(/2: [/NULL - /NULL]; tight), fd=()-->(2)]
 
 memo
 SELECT p,q,r,s FROM pqr WHERE q = 1 AND r = 1 AND s = 'foo'

--- a/pkg/sql/opt/xform/testdata/rules/select
+++ b/pkg/sql/opt/xform/testdata/rules/select
@@ -219,30 +219,32 @@ project
  ├── cardinality: [0 - 1]
  ├── key: ()
  ├── fd: ()-->(1)
- └── inner-join (zigzag a@u a@v)
+ └── select
       ├── columns: k:1(int!null) u:2(int!null) v:3(int!null)
-      ├── eq columns: [1] = [1]
-      ├── left fixed columns: [2] = [1]
-      ├── right fixed columns: [3] = [5]
       ├── cardinality: [0 - 1]
       ├── key: ()
       ├── fd: ()-->(1-3)
+      ├── scan a@v
+      │    ├── columns: k:1(int!null) u:2(int) v:3(int!null)
+      │    ├── constraint: /3: [/5 - /5]
+      │    ├── cardinality: [0 - 1]
+      │    ├── key: ()
+      │    └── fd: ()-->(1-3)
       └── filters
-           ├── u = 1 [type=bool, outer=(2), constraints=(/2: [/1 - /1]; tight), fd=()-->(2)]
-           └── v = 5 [type=bool, outer=(3), constraints=(/3: [/5 - /5]; tight), fd=()-->(3)]
+           └── u = 1 [type=bool, outer=(2), constraints=(/2: [/1 - /1]; tight), fd=()-->(2)]
 
 memo
 SELECT k FROM a WHERE u = 1 AND v = 5
 ----
-memo (optimized, ~8KB, required=[presentation: k:1])
+memo (optimized, ~6KB, required=[presentation: k:1])
  ├── G1: (project G2 G3 k)
  │    └── [presentation: k:1]
  │         ├── best: (project G2 G3 k)
- │         └── cost: 0.04
- ├── G2: (select G4 G5) (zigzag-join G5 a@u a@v) (select G6 G7) (select G8 G9)
+ │         └── cost: 1.10
+ ├── G2: (select G4 G5) (select G6 G7) (select G8 G9)
  │    └── []
- │         ├── best: (zigzag-join G5 a@u a@v)
- │         └── cost: 0.03
+ │         ├── best: (select G8 G9)
+ │         └── cost: 1.09
  ├── G3: (projections)
  ├── G4: (scan a) (scan a@u) (scan a@v)
  │    └── []

